### PR TITLE
Codechange: [OSX] Use std::unique_ptr with a custom deleter to simply…

### PR DESCRIPTION
**Pull request by michicc**
… memory management of Core Foundation types.

### CODE REVIEW

**Comment by michicc** [comment]
One thing I'm not totally happy about are all the extra `.get()`s, but eliminating them would either mean rolling a custom duplicate of `std::unique_ptr` or at least subclassing it. OTOH, there's a reason the standard does it this way to clearly indicate when ownership is transferred and when not.

File: src/os/macosx/macos.h [hunk 1]
```diff
@@ -40,4 +40,17 @@ bool IsMonospaceFont(CFStringRef name);
...
+template <typename T> struct CFDeleter {
+	void operator()(T *p)
+	{
+		if (p) ::CFRelease(p);
+	}
+};
+
+/** Specialisation of std::unique_ptr for CoreFoundation objects. */
+template <typename T>
+using CFReleaser = std::unique_ptr<typename std::remove_pointer<T>::type, CFDeleter<typename std::remove_pointer<T>::type>>;
```
Comment (nielsmh) [inline src/os/macosx/macos.h]: Suggest naming this `CFAutoResource<>` and the functor doing the release call `CFDeleter<>`.

Comment (michicc) [reply src/os/macosx/macos.h]: I'd prefer something like `CFAutoRelease<>` more, but the naming is always some matter of opinion.

Except that, is the general idea sound?

**Comment by michicc** [comment]
> Sep 02 20:49:36 <michi_cc>	andythenorth: Can you check if I broke anything with #7717 when it is convenient?
> Sep 02 20:53:04 <andythenorth>	sure :)
> Sep 02 20:59:54 <andythenorth>	michi_cc: it builds
> Sep 02 20:59:59 <andythenorth>	not sure what else to test meaningfully :)
> Sep 02 21:00:51 <andythenorth>	savegames load and run

File: src/os/macosx/macos.h (hunk 1 above)
Comment (LordAro) [reply src/os/macosx/macos.h]: Why have a named struct at all? A lambda function should do, will it not?

Comment (LordAro) [reply src/os/macosx/macos.h]: Oh, will it require a generic lambda? That needs C++17/20, iirc?

**Review by nielsmh — APPROVED** [review]

**Pull request merged**